### DISCUSSION TITLE
Is there a way to have more tightly scoped imports through project.janet?

### DISCUSSION THREAD

**Discussion by eli-oat** [janet-lang/janet#1218] [Q&A] [answered]
I know that by using the `import` macro I can include just the bits of, say, spork, that I am actually using. 

Is there a way to do this similarly through a `project.janet` file, too? 

I haven't found a way to get the `:dependencies` to scope down to a specific set of functions, macros, etc., it always grabs the entire kit and kaboodle.

**Comment by CosmicToast** [comment, marked as answer]
There isn't. How exactly do you imagine this being implemented?
Ultimately, it's all in the repository, which you must clone in its entirety either way.
You could maybe skip the build step for native components if you statically analyzed hard enough, but you're not really gaining much.
What are you looking to gain through this on the overall?
I mean, I can only really imagine namespace pollution being a reason (since due to the repo considerations earlier, you don't gain any space), but how advisable is it really to have `a/b` be "available" while depending on "a/c"?

**Reply by eli-oat** [reply]
Thanks! This is very helpful and lines up with what I assumed the answer was gonna be -- mostly I wanted to make sure I wasn't missing anything big. 

I find myself reaching regularly for a few items from spork, specifically, and it mostly just *felt* wasteful to grab the whole thing. 

In this specific case I could imagine a syntax like, 

```
(declare-project
  :name "mylib" # required
  :description "a library that does things" # some example metadata.

  # Optional urls to git repositories that contain required artifacts.
  :dependencies [{:url "https://github.com/janet-lang/spork.git"
                  :ns  "spork/json"
                  :tag "v1.1.0"}])
```

...or maybe the namespace could be an array of options? 

As you say, though, I'm not sure if there is a clear benefit to doing this, or how you'd then handle only grabbing the code necessary for that specific namespace.
